roofit/roofitcore/src/RooMinimizer.cxx

```diff
@@ -816,8 +799,61 @@ void RooMinimizer::profileStop()
 }
 
 
+ROOT::Math::IMultiGenFunction* RooMinimizer::getFitterMultiGenFcn() const
+{
+   return fitter()->GetFCN();
+}
+
+
+ROOT::Math::IMultiGenFunction* RooMinimizer::getMultiGenFcn() const
+{
+   if (getFitterMultiGenFcn()) {
+      return getFitterMultiGenFcn();
+   } else {
+      switch (_fcnMode) {
+      case FcnMode::classic: {
+         return static_cast<ROOT::Math::IMultiGenFunction *>(static_cast<RooMinimizerFcn *>(_fcn));
+      }
+      case FcnMode::gradient: {
+         auto thing = dynamic_cast<RooGradMinimizerFcn *>(_fcn);
+         if (thing != nullptr) {
+            return static_cast<ROOT::Math::IMultiGenFunction *>(thing);
+         } else {
+            throw std::logic_error("In RooMinimizer::fitterFcn: Minimizer fcnMode was set to gradient, but fit function type does not match RooGradMinimizerFcn!");
+         }
+      }
+      }
+   }
+}
 
 
+const RooAbsMinimizerFcn *RooMinimizer::fitterFcn() const
+{
+   if (getFitterMultiGenFcn()) {
+      switch (_fcnMode) {
+      case FcnMode::classic: {
+         return static_cast<RooAbsMinimizerFcn *>(static_cast<RooMinimizerFcn *>(getFitterMultiGenFcn()));
+      }
+      case FcnMode::gradient: {
+         auto thing = dynamic_cast<RooGradMinimizerFcn *>(getFitterMultiGenFcn());
+         if (thing != nullptr) {
+            return static_cast<RooAbsMinimizerFcn *>(thing);
+         } else {
+            throw std::logic_error("In RooMinimizer::fitterFcn: Minimizer fcnMode was set to gradient, but fit function type does not match RooGradMinimizerFcn!");
+         }
+      }
```

Comment:
Hmm I don't see how the switch statement can be avoided because of the two different `static_cast`, but I also get the warning about end of control flow for a non-void function. Maybe we can just return `return nullptr` at the end of the function to make the compiler happy?

Reply:
First we get rid of the nullptr and logic error check. If the compiler is smart enough, that should already stop it from complaining, since then really every block returns something. If it still complains, we can just define a `RooAbsMinimizerFcn *return_ptr` at the top, store the casted result there in each block and return at the end. What do you think?

Reply:
I just checked, but removing the logic error check doesn't suppress the warnings. I think the compiler doesn't understand that all the cases in the switch statement are exhaustive, i.e. it doesn't consider that `classic` and `gradient` are the only possible values for `_fcnMode`. Your idea with the return value sounds good to me, though!

Reply:
You have to add a `default` clause. It's because if somebody decides to add a case to the thing that you switch on, this statement wouldn't do anything --> end of non-void function.

You can even put an `assert` or `throw` in the `default` clause to make this detectable.